### ISSUE TITLE
react/no-used-prop-types error thrown when a prop is only used in nextProps

### ISSUE THREAD

**Issue by mjilugu** [yannickcr/eslint-plugin-react#1142]
When we use a prop only with nextProps to set the state in the componentWillRecieveProps(nextProps) method and not elsewhere we get `react/no-used-prop-types` error for the prop in the propTypes block.

Same as #801 

Here is an excerpt of my code:

```
class MyComponent extends React.Component {

  static propTypes = {
    isLoading: React.PropTypes.bool, // eslint-disable-line react/no-unused-prop-types
    error: React.PropTypes.bool, // eslint-disable-line react/no-unused-prop-types
  }

  static defaultProps = {
    isLoading: true,
    error: false,
  };

  constructor(props) {
    super(props);

    this.state = {
      isLoading: true,
      error: false,
    };
  }

  componentWillReceiveProps(props) {
    if (props.error) {
      this.setState({ ...this.state, error: props.error });
    } else {
      this.setState({
        ...this.state,
        isLoading: props.isLoading,
        error: false,
      });
    }
  }

  // the rest of the code
}
```
Notice I have to use `// eslint-disable-line react/no-unused-prop-types` to avoid the warnings

**Comment by borwahs** [comment]
@mjilugu & @ljharb -

Just as an FYI, I found that if you destructure `nextProps`, the linter does not throw an error in this scenario. If you use dot notation, it fails.

I also confirmed others have seen similar results in this thread: #884 (comment)

**Comment by joetidee** [comment]
Please can you fix this - having the same issue :(

**Comment by jseminck** [comment]
I could try and take a look at this. Not sure how far I will get.

I see that the same issue exists with `shouldComponentUpdate` and `nextProps`: https://github.com/yannickcr/eslint-plugin-react/issues/1213

**Comment by jseminck** [comment]
Hi,

This issue seems fixed in https://github.com/yannickcr/eslint-plugin-react/pull/1106 - E.g. it should work on the 7.0.0 release (which was released after this issue was created.).

@joetidee can you please let me know what version you use and what is failing for you?

Just to be clear, the example code from the original issue works. See below for a more concise example:

**This passes eslint-plugin-react test case so it should work**

js
```
class Hello extends Component {
  static propTypes = {
    foo: PropTypes.string
    bar: PropTypes.string
  };

  componentWillReceiveProps (nextProps) {
    if (nextProps.foo) {
      return true;
    }
  }

  render() {
    return (<div>{this.props.bar}</div>);
  }
}
```

**This does NOT pass eslint-plugin-react test case so it should NOT work**

js
```
class Hello extends Component {
  static propTypes = {
    foo: PropTypes.string
    bar: PropTypes.string
  };

  shouldComponentUpdate (nextProps) { // replace componentWIllReceiveProps with shouldComponentUpdate
    if (nextProps.foo) {
      return true;
    }
  }

  render() {
    return (<div>{this.props.bar}</div>);
  }
}
```

It seems like we only support for `nextProps` in `componentWillReceiveProps` but not inside other React lifecycle methods.

**Comment by joetidee** [comment]
Excellent - thanks - have upgraded to 7.0.0 and now fixed.

**Comment by ljharb** [comment]
#1232 hopefully has solved this; we can reopen after the next release if not.

**Comment by mqklin** [comment]
@ljharb `react/prop-types` doesn't warn if I get props from `nextProps`
![image](https://user-images.githubusercontent.com/9456433/34660032-fa5473ae-f460-11e7-967c-5575f5f1d677.png)

**Comment by ljharb** [comment]
@mqklin could you file a new issue about that?

**Comment by mqklin** [comment]
@ljharb there is already: https://github.com/yannickcr/eslint-plugin-react/issues/814

**Comment by pkhodaveissi** [comment]
still not working here
I have `componentWillReceiveProps` but I don't use the prop there, I pass nextProps to a function call inside the lifecycle method and then in there I go ahead and access the prop via dot notation or object destruction, but next to the prop declaration inside PropsType(using flow) I see `'helperAppStatus' PropType is defined but prop is never used eslint(react/no-unused-prop-types)`

**Comment by ljharb** [comment]
@pkhodaveissi if you pass a props or state object around, you’ve both defeated all static analysis and also, you’ve given up control of a mutable object that your component depends on, a bad practice.

Always extract every prop/state from the objects immediately, and pass those around.

**Comment by pkhodaveissi** [comment]
Thank you for your helpful feedback
